### PR TITLE
Adding parser and parameter files for encore

### PR DESCRIPTION
**Pull request by DSuveges**
This is a parser for the ENCORE dataset to generate disease/target evidence for significant gene pair interactions. As of Feb 14. 2022 one set of colorectal carcinoma (COLO1), and one set of the breast cancer (BRCA1) dataset is available. Also the parser only uses the GEMINI dataset for resolving significant gene interaction. We don't have the sufficient input for parsing the BLISS datasets. The data flow might significantly change depending on the answers we might get from the encore team.

**Script Input**:

* Parameter file (shipped with the script) containing general and experiment specific metadata eg. modelled disease, experiment specific confidence cutoff, original project id and description etc.
* Cell passport file: collection of all cellular models downloaded from the [Sanger model website](https://cellmodelpassports.sanger.ac.uk/downloads).

**Other files used by the script defined by the parameter file**: These are required as different experiments and analytical methods generate different datafiles.

* log fold change file: effect of each gene pair in all the studied cell lines. 
* gemini file: statistical parameters of the gene interaction in each cell lines generated by gemini method.
* bliss file: statistical parameters of the gene interaction in each cell lines generated by bliss method. (not yet used.)
 
**Some insights**:

1. Parameter file is read, experiments are processed.
2. Cell model file read and processed, diseaseCellLines object is created from all the cell models + status of microsetellite stability biomarker.
3. log fold change file is read, gemini file read, filters applied.
4. disease cell lines joined with the actual measurements.
5. Resolving gene pairs and their respective function 
6. Final formatting + saving as a gzipped json.

This parser is based on the VL release on 2021. December to get the files please contact @sarahyoung3. Sample dataset for COLO1 can be found in this bucket: `gs://ot-team/dsuveges/encore_sample_dataset`, however the parameter file needs to be updated.

### CODE REVIEW

**Review by tskir — COMMENTED** [review]
Thanks for the PR Daniel, this is obviously a major and very important update.

I've only reviewed the first half of the changes for now (general set up and cell passport preparation). I'll finish this tomorrow. However, I already have a number of comments so I'm submitting this partial review now.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:2]: That absolutely makes sense. Just give a brief description on the process should not be an issue.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:34]: You are right, I keep being inconsistent with the variable names.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:493]: Conceptually you are right. However consider this datasource: for one experiment, we use 3 different files. And we potentially have six experiments total. The same stands for the ot_crispr evidence, where there are ~15-20 different files to ingest.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:135]: Yes, I read a lot about the `case` feature of Python3.10. I'm not sure if all our dependencies are ready to move to P3.10 though.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:88]: This is a kind of temporary solution (so I hope it won't stay forever), as I'm hoping, we'll soon have a cell-line entity, and all these can be resolved by the ETL. So, I'm not sure how much energy should be invested.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:69]: Yes, exactly. For encore, this field goes directly into the evidence.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:96]: Yes, this is the case for all those evidence sources where the study was done on one single cell line. However for project score (crispr) this is a combination of experiments performed on a set of cell lines. It's ot ideal.

Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:96]: I feel we are trying to grasp too much complexity... however I see the value of having this resolution: as different cell lines have different biomarker compositions, it would (one day) allow a more granular questions.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:67]: One entry three rows:
```
SIDM01287,HCM-SANG-0290-C15,OESO_005,Organoid,Unknown,,,None,,MSS,8.24,,,False,SIDS01229,Esophagus,Tumour,Esophageal Carcinoma,>
Cisplatin
Epirubicin",Esophagus,T3,N0,M0,,Poorly differentiated,SIDP01120,Homo Sapiens,Male,White,Ex-Smoker,,,,,,HCM-SANG-0290-C15,ATCC:P>
```
from spark:
```
-RECORD 0-------------------------------------------------------
 model_id                 | SIDM01287
 model_name               | HCM-SANG-0290-C15
 synonyms                 | OESO_005
 model_type               | Organoid
 growth_properties        | Unknown
 doi                      | null
 pmed                     | null
 model_treatment          | None
 model_comments           | null
 msi_status               | MSS
 mutational_burden        | 8.24
 ploidy                   | null
 parent_id                | null
 crispr_ko_data           | False
 sample_id                | SIDS01229
 tissue                   | Esophagus
 tissue_status            | Tumour
 cancer_type              | Esophageal Carcinoma
 cancer_type_detail       | Esophageal Adenocarcinoma
 cancer_type_ncit_id      | C4025
 age_at_sampling          | 68.0
 sampling_day             | 18.0
 sampling_month           | 5.0
 sampling_year            | 2016.0
 sample_treatment         | Chemotherapy
 sample_treatment_details | Capecitabine
Cisplatin
Epirubicin
 sample_site              | Esophagus
 tnm_t                    | T3
 tnm_n                    | N0
 tnm_m                    | M0
 tnm_integrated           | null
 tumour_grade             | Poorly differentiated
 patient_id               | SIDP01120
 species                  | Homo Sapiens
 gender                   | Male
 ethnicity                | White
 smoking_status           | Ex-Smoker
 model_relations_comment  | null
 COSMIC_ID                | null
 BROAD_ID                 | null
 CCLE_ID                  | null
 RRID                     | null
 HCMI                     | HCM-SANG-0290-C15
 suppliers                | ATCC:PDM-66
 supplier                 | ATCC
 cat_number               | PDM-66
```

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py]: OK, it make sense, I'll do it.

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [reply partner_preview_scripts/encore_parser.py:375]: Agreed, I haven't really considered this. Indeed, in this case specifying the directory is the better choise. I'll mark this as resolved :)

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [reply partner_preview_scripts/encore_parser.py]: Ah, okay, I see. Just one more question though, to confirm I understand this right: it's intended that in the schema this field will allow both a single ('tissue', 'name', 'id', 'tissueId') struct, and a list of such structs?

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [inline partner_preview_scripts/encore_parser.py:324]: This is purely a matter of personal preference, but in such cases, I prefer to have a "select everything you need" approach rather than "drop everything you don't need". This also has two added bonuses:
1. Helps with readability, because at the end of a large dataframe constructions you can see which fields are being kept
2. Ensures the stable sort order of all the columns

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [inline partner_preview_scripts/encore_parser.py:79]: I know SIDM* identifiers are Sanger DepMap cancer cell line model identifiers; but what are those CSID* ones? I wasn't able to find them anywhere

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [inline partner_preview_scripts/encore_parser.py:82]: And also, again sorry if it's a stupid question, but what exactly are those values? I can see that the LFC dataset contains these values for several gene pairs in several model cell lines, but what exactly is being measured, and compared to what?

(This is the type of information I think would be extremely helpful to include in README, or even better just into the header of this module)

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py]: Fixing it.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py]: I added explanation for this.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:347]: Yes, however I would just remove this comment, no need to fix anything. This value is not shown.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:106]: I also added explanation.

File: common/evidence.py
Comment (tskir) [inline common/evidence.py:105]:
```suggestion
             # The following option is required to correctly parse CSV records which contain newline characters.
             # Example: SIDM01287
            .option('multiline', True)  
```

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [reply partner_preview_scripts/encore_parser.py]: Wow. Just then I thought there was left nothing to surprise me. And you're right, the multiline options solves this correctly. Resolving discussion

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [reply partner_preview_scripts/encore_parser.py]: I see that in the latest PR version, we're still looking up through OLS. I don't mind at all, as you correctly said it's not that much of a pain for now. However, if we're not going to do this straight away, I think it warrants a short `#TODO` in that place in the code

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [reply partner_preview_scripts/encore_parser.py]: But, and sorry if I'm being dense here, wouldn't `struct(['tissue', 'name', 'id', 'tissueId'])` just generate a single structured object rather than a list? I.e., what I think it will generate is this:

```
{'tissue': 'foo', 'name': 'bar', 'id': 123, 'tissueId': 456}
```

While the schema will expect:
```
[
    {'tissue': 'foo', 'name': 'bar', 'id': 123, 'tissueId': 456}
]
```

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [inline partner_preview_scripts/encore_parser.py:102]:
```suggestion
        # Generating struct for each cell lines:
        # SIDM are Sanger model identifiers, while CSID are replicate identifiers.
```

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [inline partner_preview_scripts/encore_parser.py:12]:
```suggestion
from functools import reduce

from pyspark.sql import SparkSession
```

(Removing extra double blank line)

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [inline partner_preview_scripts/encore_parser.py:19]:
```suggestion
from common.evidence import initialize_sparksession, write_evidence_strings, GenerateDiseaseCellLines


class EncoreEvidenceGenerator:
```

(Adding an extra blank line before class definition.)

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [inline partner_preview_scripts/encore_parser.py:45]:
```suggestion
        self.log_fold_change_cutoff_p_val = shared_parameters['logFoldChangeCutoffPVal']
        self.log_fold_change_cutoff_FDR = shared_parameters['logFoldChangeCutoffFDR']
        self.interaction_cutoff_p_val = shared_parameters['interactionCutoffPVal']
        self.data_folder = shared_parameters['data_folder']
        self.release_version = shared_parameters['releaseVersion']
        self.release_date = shared_parameters['releaseDate']
```

File: partner_preview_scripts/encore_parser.py
Comment (tskir) [inline partner_preview_scripts/encore_parser.py:53]:
```suggestion
    @udf(ArrayType(StructType([
        StructField('targetFromSourceId', StringType(), False),
        StructField('targetRole', StringType(), False),
        StructField('interactingTargetFromSourceId', StringType(), False),
        StructField('interactingTargetRole', StringType(), False),
    ])))
```

**Review by tskir — COMMENTED** [review]
Thank you for the edits Daniel, I've now finished my review. Once the remaining small comments are addressed, please feel free to merge it without further review from me.

I still have some more questions about the methodology of the underlying experiments, but we can just discuss this tomorrow privately (no need to expand the docs further)

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py]: OK, this one skipped my attention. Let's leave as it is for now. We'll get back to it later. In general I feel we should provide a very general solution for this and similar problems, which is beyond the scope of this PR.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py]: You are right, the above expression generates a single struct. I made changes to reflect this.

Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py]: Later in the code there's this line:
```python
            merged_dataset.withColumn(
                "diseaseCellLines", f.array(f.col("diseaseCellLine"))
            )
```

File: common/evidence.py
Comment (DSuveges) [reply common/evidence.py:74]: This is not an issue now, EPMC evidence generation is not in our territory.

File: partner_preview_scripts/encore_parser.py
Comment (ireneisdoomed) [inline partner_preview_scripts/encore_parser.py:428]: Does this still hold true?

File: partner_preview_scripts/encore_parser.py
Comment (ireneisdoomed) [inline partner_preview_scripts/encore_parser.py:447]:
```suggestion
```

File: partner_preview_scripts/encore_parser.py
Comment (ireneisdoomed) [inline partner_preview_scripts/encore_parser.py:482]:
```suggestion
            .select("*", f.explode(self.parse_targets(f.col("id"), f.col("Note1"))).alias("genes"))
```

File: partner_preview_scripts/encore_parser.py
Comment (ireneisdoomed) [inline partner_preview_scripts/encore_parser.py:481]:
```suggestion
```

File: partner_preview_scripts/encore_parser.py
Comment (ireneisdoomed) [inline partner_preview_scripts/encore_parser.py:485]: I'd replace the withColumns + drop with a single select

**Review by ireneisdoomed — CHANGES_REQUESTED** [review]
Congrats, parsing this data makes the LD problem look simple :)
I have comments about minor fixes, however, the logic is nicely structured so that I could follow the rationale and abstract from the complicated steps like pivoting and so on.

My main comment before merging is to fix the conflict with the `evidence` utils file and that I would prefer to have the class that parses the cell lines separated from the evidence utils. I think it is a big enough script to deserve its own location.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py]: I'm addressing this issue now.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:602]: It's been written by me, hosted in the PPP_config repository.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:129]: Yes, it would be a bit nicer, but at this point, I don't have capacity to address this point.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:309]: Yes, based on the logs, this is not an issue in the new release, but I don't want to remove the logic, because, who knows, there might be problem with it later.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:428]: Yes.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:22]: Once all the PRs are merged, I'll take a look and see what the most suitable solution would be.

File: partner_preview_scripts/encore_parser.py
Comment (DSuveges) [reply partner_preview_scripts/encore_parser.py:476]: Based on the discussion I have with Emre last year they will change the format of the files, so optimizing this implementation doesn't make much sense.

**Pull request merged**
